src/heatmaps.ts

```diff
@@ -0,0 +1,132 @@
+import { _includes, _register_event } from './utils'
+import RageClick from './extensions/rageclick'
+import { Properties } from './types'
+import { PostHog } from './posthog-core'
+
+import { document, window } from './utils/globals'
+import { getParentElement, isTag } from './autocapture-utils'
+
+type HeatmapEventBuffer =
+    | {
+          [key: string]: Properties[]
+      }
+    | undefined
+
+function elementOrParentPositionMatches(el: Element, matches: string[], breakOnElement?: Element): boolean {
+    let curEl: Element | false = el
+
+    while (curEl && !isTag(curEl, 'body')) {
+        if (curEl === breakOnElement) {
+            return false
+        }
+
+        if (_includes(matches, window?.getComputedStyle(curEl).position)) {
+            return true
+        }
+
+        curEl = getParentElement(curEl)
+    }
+
+    return false
+}
+
+export class Heatmaps {
+    instance: PostHog
+    rageclicks = new RageClick()
+    _isDisabledServerSide: boolean | null = null
+    _initialized = false
+    _mouseMoveTimeout: number | undefined
+
+    // TODO: Periodically flush this if no other event has taken care of it
+    private buffer: HeatmapEventBuffer
+
+    constructor(instance: PostHog) {
+        this.instance = instance
+
+        if (this.isEnabled) {
+            this._setupListeners()
+        }
+    }
+
+    public get isEnabled(): boolean {
+        return !!this.instance.config.__preview_heatmaps
+    }
+
+    public getAndClearBuffer(): HeatmapEventBuffer {
+        const buffer = this.buffer
+        this.buffer = undefined
+        return buffer
+    }
+
+    private _setupListeners(): void {
+        if (!window || !document) {
+            return
+        }
+
+        _register_event(document, 'click', (e) => this._onClick((e || window?.event) as MouseEvent), false, true)
+        _register_event(
+            document,
+            'mousemove',
+            (e) => this._onMouseMove((e || window?.event) as MouseEvent),
+            false,
+            true
+        )
+    }
+
+    private _getProperties(e: MouseEvent, type: string): Properties {
+        // We need to know if the target element is fixed or not
+        // If fixed then we won't account for scrolling
+        // If not then we will account for scrolling
+
+        const scrollY = this.instance.scrollManager.scrollY()
+        const scrollX = this.instance.scrollManager.scrollX()
+        const scrollElement = this.instance.scrollManager.scrollElement()
+
+        const isFixedOrSticky = elementOrParentPositionMatches(e.target as Element, ['fixed', 'sticky'], scrollElement)
```

Comment:
I wonder if we could get computed styles to avoid walking the tree but 🤷

Reply:
Sadly not. Computed styles do the whole inheritance tree but we actually don't want to know the current node's position, rather we want to know if it is in a fixed parent or not so fundamentally we have to traverse it.